### DISCUSSION TITLE
Uncertainty p-factor and r-factor with Gaussian Mixture Models (GMM)

### DISCUSSION THREAD

**Discussion by maby200** [neuralhydrology/neuralhydrology#171] [Ideas]
I am following the paper [Uncertainty estimation with deep learning for rainfall–runoff modeling](https://hess.copernicus.org/articles/26/1673/2022/) to measure uncertainty with GMM.

I am using the function [uncertainty_plot](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/evaluation/plots.py) with a slight modification to calculate **p-factor** and **r-factor**.
The definition of these factors is found on [Estimating Uncertain Flow and Transport Parameters Using a Sequential Uncertainty Fitting Procedure](https://acsess.onlinelibrary.wiley.com/doi/abs/10.2136/vzj2004.1340) (see pages 1344 for methodology and 1346-1347 for results) and in summary, it is as follows:

1. **p-factor** represents the amount of `y` inside the 2.5 and 97.5 percentiles divided by the total number of `y` observations. (code already done)
2. **r-factor** is the average width of the 2.5 and 97.5 band and the standard deviation of the observations. (currently developing the code for it)

So I think it is a good idea to implement it in the neuralhydrology package.

Any opinions?

**Comment by danklotz** [comment]
Hi Maby, 

I agree that it would be a good idea to add support for evaluating 95% prediction intervals to nh, since they are so often used (in other fields). That said, I do have some questions about your proposed metrics: 

1. Does the p-factor really consider how many samples from the estimated distribution are within the 95% prediction interval? That makes little sense to me. Are you sure you did not mean the observations `y`? 
2. Similarly I am not sure what the r-factor expresses. It seems to be a form of NSE, where the numerator is the width of the intervals instead of the squared errors. Is that correct? Is that interpretation correct? If yes, I am fine with it. 

I also have a bit concern is with the terminology that you propose (specifically the p-factor, since it does seem to just be the coverage; however, maybe also the r-factor), since I am not familiar with it and I think a lot of practitioners will not be either. From what I gather the paper you are linking to does also not use these terms. I would prefer if we go with the convention that afaik is most used in hydrology and call them reliability and resolution (as in [this paper](https://hess.copernicus.org/articles/26/1673/2022/#section2), or maybe even better use the statistical terms i.e.: coverage and interval width. With that terminology it would also become clear that the 95% interval is just one special case (which is often arbitrarily chosen as the range of interest), and you could implement a function which also allows to evaluate the metrics for different probabilities. What do you think?

**Comment by maby200** [comment]
Thank you for your response, Daniel.

Answering :
1. The actual variable to be taken into account is `y`, as you suggested. I will fix it in the original post.
2. For each observation, its resulting range of uncertainty calculated using an MDN will be averaged, i.e: (depending on which quantiles you want to study) sum each resulting range and then divide it by the number of obs. Then, this output will be divided by the standard deviation of the obs.
This r-factor is used as the ideal that the uncertainty band should be 0, i.e. the modeled values are the same as the obs and hence, no uncertainties. But as it does not always happen, it is suggested to be less than the standard deviation of the observed data.


I would like to apologize I did not specify that the p and r factor are not coined with that name in the paper I cited, but it is mentioned in [a further paper](https://www.sciencedirect.com/science/article/pii/S0022169415001985?ref=pdf_download&fr=RR-2&rr=8636a5dc0d904ffb) by the same author in the section _2.4. Calibration, parameterization, and uncertainty analysis_.
Here I cite a part of this section:
> The P-factor is the fraction of measured data bracketed by the 95% prediction uncertainty (95PPU) band and varies from 0 to 1, where 1 indicates 100% bracketing of the measured data within model prediction uncertainty (i.e., a perfect model simulation considering the uncertainty). The quantity (1 -  P factor) could hence be referred to as the model error. For discharge, we recommend a value of >0.7 or 0.75 to be adequate. This of course depends on the scale of the project and adequacy of the input and calibrating data. The R-factor on the other hand is the ratio of the average width of the 95PPU band and the standard deviation of the measured variable. A value of <1.5, again depending on the situation, would be desirable for this index (Abbaspour et al., 2004, 2007).

So as stated by the author, the intention of these factors is to inform the capabilities of a model in predicting uncertainties. Of course these factors are result of a different algorithm specific for SWAT-CUP model and, as a cosequence, may not be called the same when using MDN in LSTMs.

I agree that these names may not be used in this context.

**Reply by danklotz** [reply]
Hey, 

thank you for clarifying. Regarding the terminology: I think you can keep the naming convention given that it has been adopted by the SWAT community (as long as we explain them in general terms).

**Reply by maby200** [reply]
Great! :hugs:
